File: frontend/lib/models/apartment.dart
class Apartment {
  final String id;
  final String title;
  final String description;
  final double price;
  final String location;
  final String district;
  final int bedrooms;
  final int bathrooms;
  final double? area;
  final List<String> images;
  final List<String> amenities;
  final bool available;
  final String gender;
  final bool furnished;
  final int totalRooms;
  final int totalBeds;
  final int availableRooms;
  final int availableBeds;
  final double? pricePerRoom;
  final double? pricePerBed;
  final bool allowRoomBooking;
  final bool allowBedBooking;
  final DateTime? createdAt;

  Apartment({
    required this.id,
    required this.title,
    required this.description,
    required this.price,
    required this.location,
    required this.district,
    required this.bedrooms,
    required this.bathrooms,
    this.area,
    required this.images,
    required this.amenities,
    required this.available,
    required this.gender,
    required this.furnished,
    required this.totalRooms,
    required this.totalBeds,
    required this.availableRooms,
    required this.availableBeds,
    this.pricePerRoom,
    this.pricePerBed,
    required this.allowRoomBooking,
    required this.allowBedBooking,
    this.createdAt,
  });

  factory Apartment.fromJson(Map<String, dynamic> json) {
    return Apartment(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      location: json['location'] ?? '',
      district: json['district'] ?? '',
      bedrooms: json['bedrooms'] ?? 1,
      bathrooms: json['bathrooms'] ?? 1,
      area: json['area']?.toDouble(),
      images: (json['images'] as List?)?.cast<String>() ?? [],
      amenities: (json['amenities'] as List?)?.cast<String>() ?? [],
      available: json['available'] ?? true,
      gender: json['gender'] ?? 'any',
      furnished: json['furnished'] ?? false,
      totalRooms: json['totalRooms'] ?? 1,
      totalBeds: json['totalBeds'] ?? 2,
      availableRooms: json['availableRooms'] ?? 1,
      availableBeds: json['availableBeds'] ?? 2,
      pricePerRoom: json['pricePerRoom']?.toDouble(),
      pricePerBed: json['pricePerBed']?.toDouble(),
      allowRoomBooking: json['allowRoomBooking'] ?? false,
      allowBedBooking: json['allowBedBooking'] ?? false,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'title': title,
    'description': description,
    'price': price,
    'location': location,
    'district': district,
    'bedrooms': bedrooms,
    'bathrooms': bathrooms,
    'area': area,
    'images': images,
    'amenities': amenities,
    'available': available,
    'gender': gender,
    'furnished': furnished,
    'totalRooms': totalRooms,
    'totalBeds': totalBeds,
    'availableRooms': availableRooms,
    'availableBeds': availableBeds,
    'pricePerRoom': pricePerRoom,
    'pricePerBed': pricePerBed,
    'allowRoomBooking': allowRoomBooking,
    'allowBedBooking': allowBedBooking,
  };
}

File: backend/src/models/Apartment.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const Apartment = sequelize.define('Apartment', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  title: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: false,
  },
  price: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  location: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  district: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  bedrooms: {
    type: DataTypes.INTEGER,
    defaultValue: 1,
  },
  bathrooms: {
    type: DataTypes.INTEGER,
    defaultValue: 1,
  },
  area: {
    type: DataTypes.FLOAT,
    allowNull: true,
  },
  images: {
    type: DataTypes.TEXT,
    defaultValue: '[]',
    get() {
      const raw = this.getDataValue('images');
      try { return JSON.parse(raw); } catch { return []; }
    },
    set(val) {
      this.setDataValue('images', JSON.stringify(val));
    }
  },
  amenities: {
    type: DataTypes.TEXT,
    defaultValue: '[]',
    get() {
      const raw = this.getDataValue('amenities');
      try { return JSON.parse(raw); } catch { return []; }
    },
    set(val) {
      this.setDataValue('amenities', JSON.stringify(val));
    }
  },
  available: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
  },
  gender: {
    type: DataTypes.ENUM('any', 'male', 'female'),
    defaultValue: 'any',
  },
  furnished: {
    type: DataTypes.BOOLEAN,
    defaultValue: false,
  },
  totalRooms: {
    type: DataTypes.INTEGER,
    defaultValue: 1,
  },
  totalBeds: {
    type: DataTypes.INTEGER,
    defaultValue: 2,
  },
  availableRooms: {
    type: DataTypes.INTEGER,
    defaultValue: 1,
  },
  availableBeds: {
    type: DataTypes.INTEGER,
    defaultValue: 2,
  },
  pricePerRoom: {
    type: DataTypes.FLOAT,
    allowNull: true,
  },
  pricePerBed: {
    type: DataTypes.FLOAT,
    allowNull: true,
  },
  allowRoomBooking: {
    type: DataTypes.BOOLEAN,
    defaultValue: false,
  },
  allowBedBooking: {
    type: DataTypes.BOOLEAN,
    defaultValue: false,
  },
  roomConfig: {
    type: DataTypes.TEXT,
    defaultValue: '[]',
    get() {
      const raw = this.getDataValue('roomConfig');
      try { return JSON.parse(raw); } catch { return []; }
    },
    set(val) {
      this.setDataValue('roomConfig', JSON.stringify(val));
    }
  },
});

module.exports = Apartment;
